### PR TITLE
AC-383 - Introduce Concept download feature

### PR DESCRIPTION
**Pull request by tmarzeion**

### CODE REVIEW

File: openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/SettingsPresenterTest.java [hunk 1]
```diff
@@ -42,13 +43,15 @@
...
     private SettingsPresenter settingsPresenter;
 
     private TextView logsTV;
     @Before
     public void setUp() {
-        settingsPresenter = new SettingsPresenter(view, logger,logsTV);
+        settingsPresenter = new SettingsPresenter(view, logger,logsTV, conceptDAO);
         PowerMockito.mockStatic(OpenMRS.class);
         PowerMockito.when(OpenMRS.getInstance()).thenReturn(openMRS);
     }
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/SettingsPresenterTest.java]: What about adding test coverage  ?

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/ApplicationConstants.java [hunk 2]
```diff
@@ -80,6 +80,20 @@
...
+        public static final String STOP_DOWNLOAD_ACTION = "com.openmrs.mobile.services.conceptdownloadservice.action.stopforeground";
+    }
+
+    public abstract static class BroadcastActions {
+        public static final String BROADCAST_INTENT_ID = "com.openmrs.mobile.services.conceptdownloadservice.action.broadcastintent";
+        public static final String BROADCAST_INTENT_KEY_COUNT = "com.openmrs.mobile.services.conceptdownloadservice.broadcastintent.key.count";
+    }
+
+    public abstract static class ServiceNotificationId {
+        public static int FOREGROUND_SERVICE = 101;
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/utilities/ApplicationConstants.java]: Shouldn't this be only concept download service specific name ?

File: openmrs-client/src/main/java/org/openmrs/mobile/services/ConceptDownloadService.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                        Bitmap.createScaledBitmap(icon, 128, 128, false))
+                .setContentIntent(pendingIntent)
+                .setOngoing(true)
+                .build();
+        startForeground(ApplicationConstants.ServiceNotificationId.FOREGROUND_SERVICE,
+                notification);
+    }
+
+    private void downloadConcepts(int startIndex) {
+        RestApi service = RestServiceBuilder.createService(RestApi.class);
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/services/ConceptDownloadService.java]: Shouldn't we first check the max number of objects we can fetch form the remote  ? (if possible)

File: openmrs-client/src/main/java/org/openmrs/mobile/databases/DBOpenHelper.java [hunk 4]
```diff
@@ -64,9 +68,10 @@ public DBOpenHelper(Context context) {
     @Override
     public void onCreate(SQLiteDatabase sqLiteDatabase) {
         mLogger.d("Database creating...");
-
         sqLiteDatabase.execSQL(mPatientTable.createTableDefinition());
         logOnCreate(mPatientTable.toString());
+        sqLiteDatabase.execSQL(mConceptTable.createTableDefinition());
+        logOnCreate(mConceptTable.toString());
```
Comment (AdamGrzybkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/databases/DBOpenHelper.java]: What about onUpdate() method  ?

File: openmrs-client/src/main/java/org/openmrs/mobile/services/ConceptDownloadService.java (hunk 3 above)
Comment (tmarzeion) [inline openmrs-client/src/main/java/org/openmrs/mobile/services/ConceptDownloadService.java]: Totally forgot about that, fixed

File: openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/SettingsPresenterTest.java (hunk 1 above)
Comment (tmarzeion) [inline openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/SettingsPresenterTest.java]: Fixed

File: openmrs-client/src/main/java/org/openmrs/mobile/databases/DBOpenHelper.java (hunk 4 above)
Comment (tmarzeion) [inline openmrs-client/src/main/java/org/openmrs/mobile/databases/DBOpenHelper.java]: Done

**Comment by AvijitGhosh82** [comment]
Ping @tmarzeion

**Pull request merged**